compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+   *  filtering tests
+   */
+  private lazy val filter: Option[Regex] = regex.map(str => new Regex(str))
+
+  /** A compilation target whose files or directory of files is to be compiled
+   *  in a specific way defined by the `Test`
+   */
+  private sealed trait Target { self =>
+    def outDir: JFile
+    def flags: Array[String]
+
+    /** Adds the flags specified in `newFlags0` if they do not already exist */
+    def withFlags(newFlags0: String*) = {
+      val newFlags = newFlags0.toArray
+      if (!flags.containsSlice(newFlags)) self match {
+        case self: ConcurrentCompilationTarget =>
+          self.copy(flags = flags ++ newFlags)
+        case self: SeparateCompilationTarget =>
+          self.copy(flags = flags ++ newFlags)
+      }
+      else self
+    }
+
+    /** Generate the instructions to redo the test from the command line */
+    def buildInstructions(errors: Int, warnings: Int): String = {
+      val sb = new StringBuilder
+      val maxLen = 80
+      var lineLen = 0
+
+      sb.append(s"\n\nTest compiled with $errors error(s) and $warnings warning(s), the test can be reproduced by running:")
+      sb.append("\n\n./bin/dotc ")
+      flags.foreach { arg =>
+        if (lineLen > maxLen) {
+          sb.append(" \\\n        ")
+          lineLen = 4
+        }
+        sb.append(arg)
+        lineLen += arg.length
+        sb += ' '
+      }
+
+      self match {
+        case ConcurrentCompilationTarget(files, _, _) => {
+          files.map(_.getAbsolutePath).foreach { path =>
+            sb.append("\\\n        ")
+            sb.append(path)
+            sb += ' '
+          }
+          sb.toString + "\n\n"
+        }
+        case self: SeparateCompilationTarget => {
+          val command = sb.toString
+          val fsb = new StringBuilder(command)
+          self.compilationUnits.foreach { files =>
+            files.map(_.getPath).foreach { path =>
+              fsb.append("\\\n        ")
+              lineLen = 8
+              fsb.append(path)
+              fsb += ' '
+            }
+            fsb.append("\n\n")
+            fsb.append(command)
+          }
+          fsb.toString + "\n\n"
+        }
+      }
+    }
+  }
+
+  /** A group of files that may all be compiled together, with the same flags
+   *  and output directory
+   */
+  private final case class ConcurrentCompilationTarget(
+    files: Array[JFile],
+    flags: Array[String],
+    outDir: JFile
+  ) extends Target {
+    override def toString() = outDir.toString
+  }
+
+  /** A target whose files will be compiled separately according to their
+   *  suffix `_X`
+   */
+  private final case class SeparateCompilationTarget(
+    dir: JFile,
+    flags: Array[String],
+    outDir: JFile
+  ) extends Target {
+
+    /** Get the files grouped by `_X` as a list of groups, files missing this
+     *  suffix will be put into the same group
+     */
+    def compilationUnits: List[Array[JFile]] =
```

Comment:
Given the documentaion, I'll call this compilationGroups.
Otherwise you're re-defining a term already defined in Dotty. Every JFile here is a already a single compilationUnit.